### PR TITLE
Allow node addresses to be used as loadBalancer addresses for proxy-mode=ipvs

### LINKED ISSUE #114815: NodePort will not listen on external IP if same IP is used as loadBalancerIP
### What happened?

Hey, after upgrading kubernetes from 1.22 to 1.23, I'm experiencing strange service provisioning behavior using nodePort. I'm using kube-proxy (ipvs) + calico + metallb. 
If external IP is used by another service type (Loadbalancer - allocated by metallb), then the node owning this IP doesn't listen to NodePort on this interface.  
Unfortunately, I can't find any changes that could be causing this, any ideas?
Kubernetes 1.23
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    metallb.universe.tf/loadBalancerIPs: externalip1
  name: service
spec:
  ports:
  - name: client
    port: 31010
    protocol: TCP
    targetPort: client
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  name: service2
spec:
  ports:
  - appProtocol: http
    name: http
    nodePort: 32080
    port: 80
    protocol: TCP
    targetPort: http
  - appProtocol: https
    name: https
    nodePort: 32443
    port: 443
    protocol: TCP
    targetPort: https
  type: NodePort
```
``` console
ipvsadm -Ln | grep externalip1
node1:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
node2:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
node3:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
```
Missing externalip1 here: 
```console
ipvsadm -L -n | grep 32443
node1:
    TCP  172.17.0.1:32443 lc
    TCP  10.20.10.13:32443 lc
    TCP  10.50.135.0:32443 lc
node2:
    TCP  externalip2:32443 lc
    TCP  172.17.0.1:32443 lc
    TCP  10.20.10.12:32443 lc
    TCP  10.58.88.192:32443 lc
node3:
    TCP  externalip3:32443 lc
    TCP  172.17.0.1:32443 lc
    TCP  172.26.0.1:32443 lc
    TCP  10.20.10.10:32443 lc
    TCP  10.20.10.11:32443 lc
    TCP  10.63.32.192:32443 lc
 ```

### What did you expect to happen?

Using k8s 1.22 service with nodeport caused listening on every interface, despite using external IP by another service.

Kubernetes 1.22
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    metallb.universe.tf/loadBalancerIPs: externalip1
  name: service
spec:
  ports:
  - name: client
    port: 31010
    protocol: TCP
    targetPort: client
  type: LoadBalancer
---
apiVersion: v1
kind: Service
metadata:
  name: service2
spec:
  ports:
  - appProtocol: http
    name: http
    nodePort: 32080
    port: 80
    protocol: TCP
    targetPort: http
  - appProtocol: https
    name: https
    nodePort: 32443
    port: 443
    protocol: TCP
    targetPort: https
  type: NodePort
```
``` console
ipvsadm -Ln | grep externalip1
node1:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
node2:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
node3:
    TCP  externalip1:31010 lc
    TCP  externalip1:32010 lc
```
```console
ipvsadm -Ln | grep 32443
node1:
    TCP  172.17.0.1:32443 lc
    TCP  externalip1:32443 lc
    TCP  10.10.10.1:32443 lc
    TCP  10.36.174.0:32443 lc
node2:
    TCP  172.17.0.1:32443 lc
    TCP  externalip2:32443 lc
    TCP  10.10.10.4:32443 lc
    TCP  10.45.192.0:32443 lc
node3:
    TCP  172.17.0.1:32443 lc
    TCP  externalip3:32443 lc
    TCP  10.10.10.2:32443 lc
    TCP  10.45.11.192:32443 lc
```

### How can we reproduce it (as minimally and precisely as possible)?

Upgrade k8s from 1.22.15 to 1.23.15

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.15", GitCommit:"b84cb8ab29366daa1bba65bc67f54de2f6c34848", GitTreeState:"clean", BuildDate:"2022-12-08T10:49:13Z", GoVersion:"go1.17.13", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.15", GitCommit:"b84cb8ab29366daa1bba65bc67f54de2f6c34848", GitTreeState:"clean", BuildDate:"2022-12-08T10:42:57Z", GoVersion:"go1.17.13", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
on-premise instances
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
$ uname -a
Linux D001-FSN1DC17 6.0.8-060008-generic #202211110629-Ubuntu SMP PREEMPT_DYNAMIC Fri Nov 11 06:36:01 UTC x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>
docker
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>
calico v3.24.5

metallb v0.13.7
ipset v7.5, protocol version: 7
ipvsadm v1.31 2019/12/24 (compiled with popt and IPVS v1.2.1)

#ipvs config: 
```console
    ipvs:
      excludeCIDRs: null
      scheduler: lc
      strictARP: true
```
</details>

### PR DESCRIPTION
**Pull request by uablrek**
#### What type of PR is this?

/kind bug
/sig network
/area kube-proxy
/area ipvs

#### What this PR does / why we need it:

To use node-addresses as loadBalancerIP works with proxy-mode=iptables and did work before K8s v1.23 for proxy-mode=ipvs.

A problem with kernel versions and netlink fixed in PR https://github.com/kubernetes/kubernetes/pull/101429 caused the problem reported in #114815. The addresses assigned for NodePort became `AllAddresses - addresses(kube-ipvs0)`, but the the node-address is a loadBalancerIP it is assigned (also) to kube-ipvs0 and is thus excluded.

The security fix in https://github.com/kubernetes/kubernetes/pull/108460 prevents access to ports on the node (like ssh) to be accesible via loadBalancerIPs and clusterIPs, but if a loadBalancerIPs is also a node-ip, then **access to the node will be blocked**.

Both these problems must be taken care of in this PR.

#### Which issue(s) this PR fixes:

Fixes #114815, #117613

#### Special notes for your reviewer:

**NOTE** I am splitting this PR into one for the cleanup https://github.com/kubernetes/kubernetes/pull/115073 and will re-build this PR on top of that.

#### Does this PR introduce a user-facing change?

```release-note
Fixes a 1.26 regression in kube-proxy ipvs mode
```

### CODE REVIEW

**Comment by aojea** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115019/pull-kubernetes-e2e-gci-gce-ipvs/1614005610477522944

> Kubernetes e2e suite: [It] [sig-network] LoadBalancers should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on different nodes expand_less	


I think that IPVS is lacking some of the conntrack fixes, this test was failing before this PR

**Comment by uablrek** [comment]
> I think that IPVS is lacking some of the conntrack fixes, this test was failing before this PR

Yeah, I should probably try to compare and align ipvs with the work done for proxy-mode=iptables in this area. Is there a flaky-test issue for this?

File: pkg/proxy/ipvs/netlink.go [hunk 1]
```diff
@@ -40,4 +40,9 @@ type NetLinkHandle interface {
 	// Only the addresses of the current family are returned.
 	// IPv6 link-local and loopback addresses are excluded.
 	GetLocalAddresses(dev string) (sets.Set[string], error)
+	// GetAllLocalAddressesExcept return all local addresses on the node, except from the passed dev.
+	// This is not the same as to take the diff between GetAllLocalAddresses and GetLocalAddresses
```
Comment (aojea) [inline pkg/proxy/ipvs/netlink.go:44]: oh, subtle

File: pkg/proxy/ipvs/testing/fake_test.go [hunk 2]
```diff
@@ -23,6 +23,7 @@ import (
 	"k8s.io/kubernetes/pkg/proxy/ipvs"
 )
 
+// (I am unsure if this test has any value since it only tests the fake implementation)
```
Comment (aojea) [inline pkg/proxy/ipvs/testing/fake_test.go:26]: :)

File: pkg/proxy/ipvs/proxier.go [hunk 3]
```diff
@@ -1273,9 +1276,13 @@ func (proxier *Proxier) syncProxyRules() {
 			if proxier.ipvsScheduler == "mh" {
 				serv.Flags |= utilipvs.FlagSourceHash
 			}
-			if err := proxier.syncService(svcPortNameString, serv, true, alreadyBoundAddrs); err == nil {
+			// We must not add the address to the dummy device if it exist on another interface
```
Comment (aojea) [inline pkg/proxy/ipvs/proxier.go:1279]: I'm not familiar with the ipvs proxy, is it creating a new IP address on this dummy interface for the virtualips that doesn't exist on the node?

Comment (aojea) [reply pkg/proxy/ipvs/proxier.go:1279]: there is another dimension, GCP loadbalancers insert a route to the VIP like this
```
local 34.79.233.109 dev eth0 table local proto 66 scope host
```

```
ip route get 34.79.233.109
local 34.79.233.109 dev lo src 34.79.233.109 uid 20162
    cache <local>
```

this VIP doesn't seem to be associated to any interface
```
 $ ip add | grep 34
 $
```

@uablrek with this changes, I understand that the IPs that are in a Service and within the node will work, but what will happen with these IPs that are in the Service `{.status.loadBalancer.ingress[0].ip}` but not in an interface?

**Comment by aojea** [comment]
/lgtm cancel

one question https://github.com/kubernetes/kubernetes/pull/115019/files#r1180516307

File: pkg/proxy/ipvs/proxier.go (hunk 3 above)
Comment (aojea) [reply pkg/proxy/ipvs/proxier.go:1279]: I'm talking about this https://github.com/kubernetes/kubernetes/issues/117621, is this going to fix this or do we need an additional PR?

**Comment by aojea** [comment]
This also fixes #117613, can you please add it to the description so the bot autocloses it?

you can use a comma separated list
```
Fixes #114815,#117613
```

Also a release note may be useful too

/lgtm

**Comment by uablrek** [comment]
I was too slow. I will close https://github.com/kubernetes/kubernetes/issues/117613 with a comment that says that unless it's back ported, the fix will be in v1.28.

As for release note, I am unsure what to say, "Possibility to use node-ip as loadBalancerIP in single-node clusters is re-enabled"

**Comment by brandond** [comment]
Since this is a regression that started with 1.26, can we get it backported to 1.26 and 1.27?

**Comment by aojea** [comment]
I think it qualifies @brandond , @uablrek are you going to do the backports?

**Comment by uablrek** [comment]
I have started to check how it's done, yes. I think backport to v1.27 is simple, but v1.26 may be tricker. I am also working on https://github.com/kubernetes/kubernetes/issues/117621. Which is more urgent you think?

**Comment by aojea** [comment]
Send the one for 1.27, and a WIP for 1.26 , we'll figure out the conflicts, ... The next minor will take same time iirc

**Comment by uablrek** [comment]
Backport to v1.27 https://github.com/kubernetes/kubernetes/pull/117684

**Comment by uablrek** [comment]
Backport to v1.26 requires a backport of the related PR https://github.com/kubernetes/kubernetes/pull/115073. I tried to cherry-pick manually, and while not so complicated, there are conflicts. Once https://github.com/kubernetes/kubernetes/pull/115073 is backported, there are still minor conflicts when backporting this PR.

The conflicts are manageable, but how shall they be handled? Shall I use:
```
./hack/cherry_pick_pull.sh upstream/release-1.26 115073 115019
```
and clean up the mess?

**Comment by aojea** [comment]
> Backport to v1.26 requires a backport of the related PR #115073. I tried to cherry-pick manually, and while not so complicated, there are conflicts. Once #115073 is backported, there are still minor conflicts when backporting this PR.

This is the problems with refactors I mentioned in one of our conversations @danwinship ,  #115073 is hard to justify as backport

**Comment by uablrek** [comment]
It may be possible to do a separate PR on release-1.26.

**Comment by SoulKyu** [comment]
Hello, 
We still have the issue on our bare-metal server in 1.27.5 and 1.28.3.

If we force 1.25 everything is working fine.

Did we miss something ?
BR

**Comment by uablrek** [comment]
Please see also https://github.com/kubernetes-sigs/kubespray/issues/10572#issuecomment-1785451612

**Pull request merged**